components/tidb_query/src/rpn_expr/impl_math.rs

```diff
@@ -347,6 +351,26 @@ fn pow(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<Real>> {
     }
 }
 
+#[inline]
+#[rpn_fn(capture = [metadata], metadata_ctor = init_rng_data)]
+fn rand(metadata: &RefCell<MySQLRng>) -> Result<Option<Real>> {
+    let mut rng = metadata.borrow_mut();
+    let res = rng.gen();
+    Ok(Real::new(res).ok())
+}
+
+fn init_rng_data(_expr: &mut Expr) -> Result<RefCell<MySQLRng>> {
+    Ok(RefCell::new(MySQLRng::new()))
+}
+
+#[inline]
+#[rpn_fn]
+fn rand_with_seed(seed: &Option<i64>) -> Result<Option<Real>> {
+    let mut rng = MySQLRng::new_with_seed(seed.clone().unwrap_or(0));
```

Comment:
looks like you can use `seed.copied()`

Reply:
do you mean `seed.cloned()?

Reply:
I think we can't use copied or cloned because it seem rnp_fn only support &Option<T> and the trait bounds were not satisfied.
Or is there another way?

Reply:
Your seed is `&Option<i64>` (which is `Copy`able) so that `seed.copied()` (instead of your current `clone()`) should work for better performance.

Reply:
I mean the copied function only for `Option<&T>` and `Option<& mut T>` and seem rpn_fn only support `&[&Option<T>]` or `&Option<T>`.

Reply:
so can we just `seed.unwrap_or(0)`?

Reply:
refactored.